todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskViewModel.java

```diff
@@ -0,0 +1,97 @@
+package com.example.android.architecture.blueprints.todoapp.addedittask;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.example.android.architecture.blueprints.todoapp.data.Task;
+import com.example.android.architecture.blueprints.todoapp.data.source.TasksDataSource;
+
+import java.util.concurrent.Callable;
+
+import rx.Completable;
+import rx.Observable;
+import rx.functions.Action0;
+import rx.functions.Func1;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Listens to user actions from the UI ({@link AddEditTaskFragment}), retrieves the task from
+ * repository, creates and updates a task.
+ */
+public class AddEditTaskViewModel {
+
+    @NonNull
+    private final TasksDataSource mTasksRepository;
+
+    @Nullable
+    private final String mTaskId;
+
+    /**
+     * Creates a ViewModel for the add/edit view.
+     *
+     * @param taskId          ID of the task to edit or null for a new task
+     * @param tasksRepository a repository of data for tasks
+     */
+    public AddEditTaskViewModel(@Nullable String taskId,
+                                @NonNull TasksDataSource tasksRepository) {
+        mTaskId = taskId;
+        mTasksRepository = checkNotNull(tasksRepository, "tasksRepository cannot be null!");
+    }
+
+    @NonNull
+    private void saveTask(@NonNull Task newTask) throws Exception {
+        if (newTask.isEmpty()) {
+            throw new Exception("Trying to save an empty task");
```

Comment:
Exception is not the pretty way of doing, this, use maybe `RuntimeException`.
Also, you will be able to remove `throws Exception`